inference-engine/src/mkldnn_plugin/nodes/mkldnn_topk_node.cpp

```diff
@@ -40,190 +1041,541 @@ bool MKLDNNTopKNode::isSupportedOperation(const std::shared_ptr<ngraph::Node>& o
+                    // empty tail
+                    for (int j = 0; sub_start + j + minor_len < n && j < minor_len; j++) {
+                        buf[cnt++] = (sub_start + j) * sort_stride;
+                        buf[cnt++] = (sub_start + j + minor_len) * sort_stride;
+                    }
+                }
+            }
+        }
+    }
+
+    // last round sort
+    int sub_p = p >> 1;
+    for (int i = sub_p - 1; p - i - 1 < n && i >= 0; i--) {
+        buf[cnt++] = i * sort_stride;
+        buf[cnt++] = (p - i - 1) * sort_stride;
+    }
+    for (; sub_p > 0; sub_p >>= 1) {
+        // support partial sort as well as full sort
+        for (int sub_start = 0; (!cmp_val || (cmp_val && sub_start < n)) && sub_start < p;
+             sub_start += sub_p) {
+            int minor_p = sub_p >> 1;
+            for (int j = 0; sub_start + j + minor_p < n && j < minor_p; j++) {
+                buf[cnt++] = (sub_start + j) * sort_stride;
+                buf[cnt++] = (sub_start + j + minor_p) * sort_stride;
+            }
+        }
+    }
+}
 
-    if (src_k == 1) {
+void MKLDNNTopKNode::topk_ref(const float *in_ptr, float *out_ptr, int32_t *dst_idx) {
+    if (top_k == 1) {
         if (is_last_dim) {
             if (mode_max)
-                top1<std::greater>(src, dst_data, dst_idx, in_dims);
+                top1(in_ptr, out_ptr, dst_idx, src_dims, [](float x, float y)->float { return x > y; });
             else
-                top1<std::less>(src, dst_data, dst_idx, in_dims);
+                top1(in_ptr, out_ptr, dst_idx, src_dims, [](float x, float y)->float { return x < y; });
         } else {
             if (mode_max)
-                top1_axis<cmpgt_ps, std::greater>(src, dst_data, dst_idx, in_dims);
+                top1_axis(in_ptr, out_ptr, dst_idx, src_dims, [](float x, float y)->float { return x > y; });
             else
-                top1_axis<cmplt_ps, std::less>(src, dst_data, dst_idx, in_dims);
+                top1_axis(in_ptr, out_ptr, dst_idx, src_dims, [](float x, float y)->float { return x < y; });
         }
     } else {
         if (is_last_dim) {
             if (mode_max)
-                topk<std::greater>(src, dst_data, dst_idx, in_dims);
+                topk(in_ptr, out_ptr, dst_idx, src_dims, [](float x, float y)->float { return x > y; });
             else
-                topk<std::less>(src, dst_data, dst_idx, in_dims);
+                topk(in_ptr, out_ptr, dst_idx, src_dims, [](float x, float y)->float { return x < y; });
         } else {
             if (mode_max)
-                topk_axis<cmpgt_ps, std::greater>(src, dst_data, dst_idx, in_dims);
+                topk_axis(in_ptr, out_ptr, dst_idx, src_dims, [](float x, float y)->float { return x > y; });
             else
-                topk_axis<cmplt_ps, std::less>(src, dst_data, dst_idx, in_dims);
+                topk_axis(in_ptr, out_ptr, dst_idx, src_dims, [](float x, float y)->float { return x < y; });
```

Comment:
Why do we need all these top1, topk and topk_axis customizations for reference. Could you please leave for reference as simple implementation as possible (reusing pre-calculated O, A and I sizes).

Reply:
Fixed. Thanks Dmitry!